File: src/state.py
import random
import math

class State:
    def __init__(self, n=15, ncolors=5, limit=3):
        self.n = n
        self.ncolors = ncolors
        self.limit = limit
        self.highlighted_count = math.ceil(self.n / 2)
        self.reset(starting=True)

    def reset(self, starting = False):
        self.clicks = [0] * self.n
        self.score = 0
        self.mode = "pass"
        self.highlighted_color = 0  # Hard-coded highlighted color
        self.clicks = [0] * self.n
        available_colors = min(self.limit*2-3, self.ncolors) # get how many colors to generate
        if starting:
            self.cells = [self.highlighted_color] * self.highlighted_count + [self.ncolors - 1 - self.highlighted_color] * (self.n - self.highlighted_count)
        else:
            self.cells = [self.highlighted_color] * self.highlighted_count + \
                        [random.choice([i for i in range(available_colors) if i != self.highlighted_color]) for _ in range(self.n - self.highlighted_count)]
            random.shuffle(self.cells)
        self._pass_()

    def is_game_over(self):
        return any(click >= self.limit for click in self.clicks)

    def get_game_state(self):
        return list(self.cells), list(self.clicks), self.score

    def set_game_state(self, state):
        cells, clicks, score = state
        self.cells, self.clicks, self.score = list(cells), list(clicks), score

    def sort(self):
        index_lists = [[] for _ in range(self.ncolors)]
        for index, color_code in enumerate(self.cells):
            index_lists[color_code].append(index)
        indicies = [flattened for index_list in index_lists for flattened in index_list]
        def index_sort(array):
            return [array[i] for i in indicies]
        
        self.cells = index_sort(self.cells)
        self.clicks = index_sort(self.clicks)
        cells, clicks, score = self.latest_pass
        self.latest_pass = index_sort(cells), index_sort(clicks), score

    def get_palette_color_count(self):
        counts = [0] * self.ncolors
        for color_code in self.cells:
            counts[color_code] += 1
        return counts

    def get_mode(self):
        if self.cells.count(self.highlighted_color) != 0: return "edit"

        if max(self.get_palette_color_count()) == self.highlighted_count:
            return "pass"
        else:
            return "fail"

    def _pass_(self):
        if all(click >= 1 for click in self.clicks):
            self.clicks = [click - 1 for click in self.clicks]
            self.score = self.score + 1
        color_counts = self.get_palette_color_count()
        self.highlighted_color = color_counts.index(self.highlighted_count)
        self.latest_pass = self.get_game_state()

    def _fail_(self):
        self.set_game_state(self.latest_pass)

    def click(self, index, color_code):
        if color_code == self.highlighted_color: return
        if self.is_game_over(): return
        if index >= self.n or color_code >= self.ncolors: return
        if self.cells[index] != self.highlighted_color: return
        self.cells[index] = color_code
        self.clicks[index] += 1
        mode = self.get_mode()
        if mode == "pass":
            self._pass_()
        if mode == "fail":
            self._fail_()
        self.mode = mode

File: app/Guyner/components/State.ts
import { useState } from 'react';

const shuffleArray = (array: any[]) => {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
};

class State {
  n: number;
  ncolors: number;
  limit: number;
  highlighted_count: number;
  cells: number[];
  clicks: number[];
  score: number;
  mode: string;
  highlightedColor: number;
  latestPass: { cells: number[]; clicks: number[]; score: number };

  constructor(n: number = 15, ncolors: number = 5, limit: number = 3) {
    this.n = n;
    this.ncolors = ncolors;
    this.limit = limit;
    this.highlighted_count = Math.ceil(n / 2);
    this.cells = [];
    this.clicks = [];
    this.score = 0;
    this.mode = 'edit';
    this.highlightedColor = 0;
    this.latestPass = { cells: [], clicks: [], score: 0 };
    this.reset(true);
  }

  reset(starting: boolean = false) {
    const available_colors = Math.min(this.limit * 2 - 3, this.ncolors);
    if (starting) {
      this.cells = Array(this.highlighted_count).fill(this.highlightedColor).concat(
        Array(this.n - this.highlighted_count).fill(this.ncolors - 1 - this.highlightedColor)
      );
    } else {
      this.cells = Array(this.highlighted_count).fill(this.highlightedColor).concat(
        Array(this.n - this.highlighted_count).fill(0).map(() => {
          let color;
          do {
            color = Math.floor(Math.random() * available_colors);
          } while (color === this.highlightedColor);
          return color;
        })
      );
      this.cells = shuffleArray(this.cells);
    }
    this.clicks = Array(this.n).fill(0);
    this.score = 0;
    this.mode = 'pass';
    this._pass_();
  }

  is_game_over() {
    return this.clicks.some(click => click >= this.limit);
  }

  get_game_state() {
    return { cells: [...this.cells], clicks: [...this.clicks], score: this.score };
  }

  set_game_state(state: { cells: number[]; clicks: number[]; score: number }) {
    this.cells = [...state.cells];
    this.clicks = [...state.clicks];
    this.score = state.score;
  }

  sort() {
    const index_lists = Array.from({ length: this.ncolors }, () => []);
    this.cells.forEach((color_code, index) => index_lists[color_code].push(index));

    const indices = index_lists.flat();
    const index_sort = (array: number[]) => indices.map(i => array[i]);

    this.cells = index_sort(this.cells);
    this.clicks = index_sort(this.clicks);
    const { cells, clicks, score } = this.latestPass;
    this.latestPass = { cells: index_sort(cells), clicks: index_sort(clicks), score };
  }

  get_palette_color_count() {
    const counts = Array(this.ncolors).fill(0);
    this.cells.forEach(color_code => counts[color_code]++);
    return counts;
  }

  get_mode() {
    if (this.cells.includes(this.highlightedColor)) return 'edit';
    return Math.max(...this.get_palette_color_count()) === this.highlighted_count ? 'pass' : 'fail';
  }

  _pass_() {
    if (this.clicks.every(click => click >= 1)) {
      this.clicks = this.clicks.map(click => click - 1);
      this.score++;
    }
    const color_counts = this.get_palette_color_count();
    this.highlightedColor = color_counts.indexOf(this.highlighted_count);
    this.latestPass = this.get_game_state();
  }

  _fail_() {
    this.set_game_state(this.latestPass);
  }

  click(index: number, color_code: number) {
    if (color_code === this.highlightedColor) return;
    if (this.is_game_over()) return;
    if (index >= this.n || color_code >= this.ncolors) return;
    if (this.cells[index] !== this.highlightedColor) return;
    this.cells[index] = color_code;
    this.clicks[index]++;
    this.mode = this.get_mode();
    if (this.mode === 'pass') this._pass_();
    if (this.mode === 'fail') this._fail_();
  }
}

export default State;
